src/runtime/heap/mem_debug.c

```diff
@@ -0,0 +1,199 @@
+ * creation, where 0 uses the default value. Use a non-zero padding if the heap
+ * is expected to return aligned values. The backed_heap version of this wrapper
+ * defaults to PAGESIZE padding.
+ * Additional checks can be performed if defined:
+ * MEMDBG_OVERRUN sets a pattern in the padding around the requested allocation
+ * that is checked on deallocation to see if it has been modified, indicating
+ * an overrun (0xcafefade).
+ * MEMDBG_INIT sets a pattern on the requested allocation to try to trigger a
+ * fault is something uses the memory without initialization (0xfeedbeef).
+ * MEMDBG_FREE sets a pattern on deallocation for the entire allocation to try
+ * to catch any use after free (0xdeaddead).
+ */
+#define MEMDBG_OVERRUN
+#define MEMDBG_INIT
+#define MEMDBG_FREE
+
+#define DBG_HDR_SIG 0xd00d00d00d00d00d
+#define PAD_MIN 64
+#define PAD_MIN_BACKED PAGESIZE
+
+typedef struct mem_debug_heap {
+    struct heap h;
+    heap parent;
+    u64 padsize;
+} *mem_debug_heap;
+
+#ifdef KERNEL
+typedef struct mem_debug_backed_heap {
+    struct backed_heap bh;
+    backed_heap parent;
+    u64 padsize;
+} *mem_debug_backed_heap;
+#endif
+
+typedef struct mem_debug_hdr {
+    u64 sig;
+    u64 allocsize;
+    u64 padsize;
+    u64 alloc_addr;
+} *mem_debug_hdr;
+
+u32 pat_init = 0xfeedbeef;
+u32 pat_freed = 0xdeaddead;
+u32 pat_redzone = 0xcafefade;
+
+#if defined(MEMDBG_OVERRUN) || defined(MEMDBG_INIT) || defined(MEMDBG_FREE)
+static void set_pattern(void *v, bytes sz, void *p, bytes psz)
+{
+    u8 *bp = v;
+    for (s64 ssz = sz; ssz > 0; ssz -= psz, bp += psz)
+        runtime_memcpy(bp, p, MIN(psz, ssz));
+}
+
+static boolean check_pattern(void *v, bytes sz, void *p, bytes psz)
+{
+    u8 *bp = v;
+    for (s64 ssz = sz; ssz > 0; ssz -= psz, bp += psz)
+        if (runtime_memcmp(bp, p, MIN(psz, ssz)) != 0)
+            return false;
+    return true;
+}
+#endif
+
+static void get_debug_alloc_size(bytes b, bytes padsize, bytes *nb, bytes *padding)
+{
+    *padding = MAX(padsize, PAD_MIN);
+    *nb = *padding * 2 + b;
+    if (padsize >= PAGESIZE)
+        *nb = pad(*nb, PAGESIZE);
+}
+
+/* These functions use volatile so the hdr address won't be optimized out when debugging. */
+static u64 alloc_check(volatile mem_debug_hdr hdr, bytes b, bytes padding)
+{
+    #ifdef MEMDBG_OVERRUN
+    set_pattern(hdr, padding, &pat_redzone, sizeof(pat_redzone));
+    set_pattern((u8 *)hdr + padding + b, padding, &pat_redzone, sizeof(pat_redzone));
+    #endif
+    hdr->sig = DBG_HDR_SIG;
+    hdr->allocsize = b;
+    hdr->padsize = padding;
+    u8 *buf = (u8 *)hdr + padding;
+    #ifdef MEMDBG_INIT
+    set_pattern(buf, b, &pat_init, sizeof(pat_init));
+    #endif
+    return u64_from_pointer(buf);
+}
+
+static void dealloc_check(volatile mem_debug_hdr hdr, u64 a, bytes b, bytes nb, bytes padding)
+{
+    assert(hdr->sig == DBG_HDR_SIG);
+    assert(b == hdr->allocsize);
+    #ifdef MEMDBG_OVERRUN
```

Comment:
Just an aesthetic nit, but in general we haven't been indenting preprocessor directives in the source.